### PR TITLE
Add support for AccessToken_With_AuthScheme cache type

### PR DESCRIPTION
**Pull request by hectormmg**
This PR:
- Extends `AccessTokenEntity` class to differentiate between Bearer and PoP (with Auth Scheme) access tokens in the cache.
- Enables the storage of a Bearer and PoP token for the same resource simultaneously
- Fixes a bug where Bearer tokens were being requested when refreshing a PoP token.

### CODE REVIEW

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 1]
```diff
@@ -1100,6 +1100,12 @@ export abstract class ClientApplication {
         const scopes = [...((request && request.scopes) || [])];
         const correlationId = (request && request.correlationId) || this.browserCrypto.createNewGuid();
 
+        // Set authenticationScheme to BEARER if not explicitly set in the request
+        if (!request.authenticationScheme) {
+            request.authenticationScheme = AuthenticationScheme.BEARER;
+            this.logger.verbose("Authentication Scheme wasn't explicitly set in request, defaulting to \"Bearer\" request");
+        }
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/ClientApplication.ts:1107]: Can you please add verbose log messages for each side of this conditional?

File: lib/msal-common/src/cache/CacheManager.ts [hunk 2]
```diff
@@ -760,7 +766,10 @@ export abstract class CacheManager implements ICacheManager {
      * @param credentialType
      */
     private matchCredentialType(entity: CredentialEntity, credentialType: string): boolean {
-        return (entity.credentialType && credentialType.toLowerCase() === entity.credentialType.toLowerCase());
+        if (entity.credentialType && credentialType.toLowerCase() === entity.credentialType.toLowerCase()) {
+            return true;
+        }
+        return false;
```
Comment (jasonnutter) [inline lib/msal-common/src/cache/CacheManager.ts:772]: I don't think this change is needed unless you want to add log messages (probably not needed).

**Review by jasonnutter — APPROVED** [review]
A couple comments, otherwise looks good!

**Review by sameerag — CHANGES_REQUESTED** [review]
Lgtm. 

Can we also test this for msal-node to ensure current usage is not broken? The authentication scheme should default to `Bearer` in the cache and adding test cases there for node should help. I will approve after we confirm node works without breaking.

File: lib/msal-common/src/cache/CacheManager.ts (hunk 2 above)
Comment (hectormmg) [reply lib/msal-common/src/cache/CacheManager.ts:772]: Right, made this change to debug, reverting.

File: lib/msal-common/src/cache/entities/AccessTokenEntity.ts [hunk 3]
```diff
@@ -88,6 +93,19 @@ export class AccessTokenEntity extends CredentialEntity {
...
+        // Create Access Token With AuthScheme instead of regular access token
+        if (atEntity.tokenType === AuthenticationScheme.POP) {
+            atEntity.credentialType = CredentialType.ACCESS_TOKEN_WITH_AUTH_SCHEME;
+            // Make sure keyId is present and add it to credential
+            const tokenClaims: TokenClaims | null = AuthToken.extractTokenClaims(accessToken, cryptoUtils);
+            if (!tokenClaims?.cnf?.kid) {
+                throw ClientAuthError.createTokenClaimsRequiredError();
+            } else {
+                atEntity.keyId = tokenClaims.cnf.kid;
+            }
```
Comment (pkanher617) [inline lib/msal-common/src/cache/entities/AccessTokenEntity.ts:106]: don't need `else` here

File: lib/msal-common/src/cache/CacheManager.ts [hunk 4]
```diff
@@ -575,10 +578,10 @@ export abstract class CacheManager implements ICacheManager {
      * @param scopes
      * @param environment
      */
-    readCacheRecord(account: AccountInfo, clientId: string, scopes: ScopeSet, environment: string): CacheRecord {
+    readCacheRecord(account: AccountInfo, clientId: string, scopes: ScopeSet, environment: string, authScheme: AuthenticationScheme): CacheRecord {
```
Comment (pkanher617) [inline lib/msal-common/src/cache/CacheManager.ts:581]: Update typedocs

File: lib/msal-common/src/cache/CacheManager.ts [hunk 5]
```diff
@@ -639,17 +642,20 @@ export abstract class CacheManager implements ICacheManager {
      * @param scopes
      * @param inputRealm
      */
-    readAccessTokenFromCache(clientId: string, account: AccountInfo, scopes: ScopeSet): AccessTokenEntity | null {
+    readAccessTokenFromCache(clientId: string, account: AccountInfo, scopes: ScopeSet, authScheme: AuthenticationScheme): AccessTokenEntity | null {
```
Comment (pkanher617) [inline lib/msal-common/src/cache/CacheManager.ts:645]: update typedocs

File: lib/msal-common/src/cache/entities/CredentialEntity.ts [hunk 6]
```diff
@@ -101,6 +101,9 @@ export class CredentialEntity {
      */
     static getCredentialType(key: string): string {
         if (key.indexOf(CredentialType.ACCESS_TOKEN.toLowerCase()) !== -1) {
+            if (key.indexOf(CredentialType.ACCESS_TOKEN_WITH_AUTH_SCHEME.toLowerCase()) !== -1) {
+                return CredentialType.ACCESS_TOKEN_WITH_AUTH_SCHEME;
+            }
```
Comment (pkanher617) [inline lib/msal-common/src/cache/entities/CredentialEntity.ts:106]: Might be good to add a comment explaining this new credential type?

**Review by pkanher617 — APPROVED** [review]
Couple small comments, also make sure we add some tests. Otherwise looks great.

File: lib/msal-common/src/cache/CacheManager.ts (hunk 4 above)
Comment (hectormmg) [reply lib/msal-common/src/cache/CacheManager.ts:582]: Done

File: lib/msal-common/src/cache/CacheManager.ts (hunk 5 above)
Comment (hectormmg) [reply lib/msal-common/src/cache/CacheManager.ts:646]: Done

**Review by sameerag — APPROVED** [review]
Thanks for adding the check for node. Can we also check

**Review by pkanher617 — APPROVED** [review]

**Pull request merged**